finstmt/inc/data.py

```diff
@@ -1,43 +1,85 @@
 from finstmt.inc.config import INCOME_STATEMENT_INPUT_ITEMS
 
 
 @dataclass(unsafe_hash=True)
 class IncomeStatementData(FinDataBase):
-    revenue: Optional[float] = 0
-    cogs: Optional[float] = 0
-    sga: Optional[float] = 0
-    int_exp: Optional[float] = 0
-    tax_exp: Optional[float] = 0
-    rd_exp: Optional[float] = 0
-    dep_exp: Optional[float] = 0
-    other_op_exp: Optional[float] = 0
-    gain_on_sale_invest: Optional[float] = 0
-    gain_on_sale_asset: Optional[float] = 0
-    impairment: Optional[float] = 0
-
-    op_exp: Optional[float] = None
-    ebit: Optional[float] = None
-    ebt: Optional[float] = None
-    net_income: Optional[float] = None
-
-    items_config = INCOME_STATEMENT_INPUT_ITEMS
-
-    @property
-    def gross_profit(self) -> Optional[float]:
-        if self.revenue is None or self.cogs is None:
-            return None
-        return self.revenue - self.cogs
-
-    @property
-    def effective_tax_rate(self) -> float:
-        if self.ebt is None:
-            raise ValueError("cannot calculate effective tax rate as ebt is None")
-        elif self.tax_exp is None:
-            raise ValueError(
-                "cannot calculate effective tax rate is tax expense is None"
+    t = symbols("t", cls=Idx)
+
+    def __init__(self, *args, **kwargs):
+        _fields = [
+            (
+                item.key,
+                numpy.float64,
+                field(default=0, repr=(False if item.key == "gross_profit" else True)),
             )
-        return self.tax_exp / self.ebt
+            for item in self.items_config_list
+        ]
+        sort_order = {
+            "revenue": 0,
+            "cogs": 1,
+            "gross_profit": 2,
+            "sga": 3,
+            "int_exp": 4,
+            "tax_exp": 5,
+            "rd_exp": 6,
+            "dep_exp": 7,
+            "other_op_exp": 8,
+            "gain_on_sale_invest": 9,
+            "gain_on_sale_asset": 10,
+            "impairment": 11,
+            "op_exp": 12,
+            "ebit": 13,
+            "ebt": 14,
+            "net_income": 15,
+        }
+        _fields.sort(key=lambda tup: sort_order[tup[0]])
+        MyClass = make_dataclass(
+            "IncomeStatementData", fields=_fields, bases=(FinDataBase,)
+        )
+        MyClass.__module__ = "finstmt.inc.data"
+        self.__class__ = MyClass
+
+        for key, value in kwargs.items():
+            # print(f"{key}: {value}")
+            setattr(self, key, value)
+
+        self.items_config = DataConfigManager(deepcopy(self.items_config))
+
+        for item in self.items_config:
+            if item.force_positive and item.extract_names is not None:
+                # If extracted and need to force positive, take absolute value
+                value = getattr(self, item.key)
+                if value is None:
+                    continue
+                positive_value = abs(value)
+                setattr(self, item.key, positive_value)
+
```

Comment:
We won't need to repeat this logic if we move away from this static class + dynamic reassignment structure.

Reply:
class deleted and moved logic to `FinDataBase`